nemo/collections/asr/modules/conformer_encoder.py

```diff
@@ -0,0 +1,303 @@
+
+        if self_attention_model == "rel_pos":
+            self.pos_enc = RelPositionalEncoding(
+                d_model=d_model, dropout_rate=dropout, dropout_emb_rate=dropout_emb, xscale=self.xscale
+            )
+        elif self_attention_model == "abs_pos":
+            self.pos_enc = PositionalEncoding(
+                d_model=d_model, dropout_rate=dropout, max_len=6000, reverse=False, xscale=self.xscale
+            )
+        else:
+            raise ValueError(f"Not valid self_attention_model: '{self_attention_model}'!")
+
+        self.layers = nn.ModuleList()
+        for i in range(n_layers):
+            layer = ConformerEncoderBlock(
+                d_model=d_model,
+                d_ff=d_ff,
+                conv_kernel_size=conv_kernel_size,
+                self_attention_model=self_attention_model,
+                n_heads=n_heads,
+                dropout=dropout,
+                dropout_att=dropout_att,
+            )
+            self.layers.append(layer)
+
+        if feat_out > 0 and feat_out != self.output_dim:
+            self.out_proj = nn.Linear(self.feat_out, feat_out)
+            self._feat_out = feat_out
+        else:
+            self.out_proj = None
+            self._feat_out = d_model
+
+    @typecheck()
+    def forward(self, audio_signal, length):
+        audio_signal = torch.transpose(audio_signal, 1, 2)
+
+        if isinstance(self.pre_encode, ConvSubsampling):
+            audio_signal, length = self.pre_encode(audio_signal, length)
+        else:
+            audio_signal = self.embed(audio_signal)
+
+        audio_signal, pos_emb = self.pos_enc(audio_signal)
+        bs, xmax, idim = audio_signal.size()
+
+        # Create the self-attention and padding masks
+        pad_mask = self.make_pad_mask(length, max_time=xmax, device=audio_signal.device)
+        xx_mask = pad_mask.unsqueeze(1).repeat([1, xmax, 1])
+        xx_mask = xx_mask & xx_mask.transpose(1, 2)
+        pad_mask = (~pad_mask).unsqueeze(2)
+
+        for lth, layer in enumerate(self.layers):
+            audio_signal = layer(x=audio_signal, att_mask=xx_mask, pos_emb=pos_emb, pad_mask=pad_mask,)
+
+        if self.out_proj is not None:
+            audio_signal = self.out_proj(audio_signal)
+
+        audio_signal = torch.transpose(audio_signal, 1, 2)
+        return audio_signal, length
+
+    @staticmethod
+    def make_pad_mask(seq_lens, max_time, device=None):
+        """Make masking for padding."""
+        bs = seq_lens.size(0)
+        seq_range = torch.arange(0, max_time, dtype=torch.int32)
+        seq_range_expand = seq_range.unsqueeze(0).expand(bs, max_time)
+        seq_lens = seq_lens.type(seq_range_expand.dtype).to(seq_range_expand.device)
+        seq_length_expand = seq_lens.unsqueeze(-1)
+        mask = seq_range_expand < seq_length_expand
+
+        if device:
+            mask = mask.to(device)
+        return mask
+
+
+class ConformerEncoderBlock(torch.nn.Module):
```

Comment:
I would urge you to subclass NeuralModule and add type checking. Alternatively, please move this to parts as opposed to keeping it inside the modules directory.

Reply:
Moved.

Reply:
Can you move this one too?

Reply:
moved.